### ISSUE TITLE
EPIC: Dual staking

### ISSUE THREAD

**Issue by lukasz-zimnoch** [mezo-org/mezod#116]
---

⚠️ This EPIC was written based on the [Dual Staking (core)](https://coda.io/d/_d_vIDNPK008#Draft-Product-Pitches_tuH_B/r29) product pitch which is the only source of truth for the dual staking feature. Please double-check the requirements before starting the work on tasks being part of this epic. Any inaccuracies should be resolved in favor of the product pitch and epic items should be updated accordingly.

---

This epic captures the work necessary to implement the Mezo dual staking mechanism. The proposed mechanism's architecture (on the Cosmos SDK module level) is presented on the following chart ([FigJam source](https://www.figma.com/board/Ms2uESVOnP6qDx0hqa4MT7/Chain-architecture?node-id=62-566&t=EwFCiRskL9PaVIKN-4)):

<img width="4960" alt="dual-staking" src="https://github.com/thesis/mezo/assets/11180469/3d17f661-b12e-47ec-8b2b-b6303d8bb293">

### Involved modules
- `x/evm`: Existing Evmos module
- `x/bank`: Existing Cosmos SDK module
- `x/staking`: Existing Cosmos SDK module
- `x/dualstaking`: Planned Mezo module

### Assumptions
- Staking tokens, validator-escrowed (VE) tokens, and staking contract will be implemented as EVM precompiles. Those contracts will be exposed on predefined addresses and `x/evm` module will be intercepting transactions and calls coming there. The actual logic will be performed by Go code sitting at the intersection of `x/evm` and relevant modules. End user interaction will be done through regular EVM transactions.
- The `x/staking` module will be modified to accept SSC (synthetic staking coin) as stake coin. Original staking tokens (BTC and MEZO) will be exchanged for SSC. Moreover, the `x/staking` module will not handle regular Cosmos messages. Staking and unstaking will happen through EVM transactions exclusively.

### Flow
1. A staker willing to stake BTC or MEZO calls the precompiled staking contract (EVM transaction) to lock their token for a specific locktime. Unstaking won't be available until the locktime expires. As part of the transaction, the staker chooses a validator for stake delegation. The staker can also decide to postpone delegation and obtain VE tokens directly. The staker may also decide to self-delegate and become a validator on their own. This step will modify the state of `x/dualstaking` module - information about the stake will be registered using module's keeper.
2.  The `x/dualstaking` module calls the `x/bank` module and transfers the staked token from the stake owner to their own module account. This puts the stake under module's escrow. The `x/dualstaking` module mints a corresponding amount of the VE token in return.
3. The `x/dualstaking` module escrows the minted VE token, calculates the amount of SSC to mint, and calls the `x/bank` module to do so. Minted coins are assigned to the validator's account. 
4. The `x/dualstaking` module calls `x/staking` module to stake and delegate minted SSC. In case of self-delegation, this step creates the validator first (if that turns out to be problematic here, we may consider to extract validator creation to a separate staking contract transaction).
5. The `x/stake` module calls `x/bank` to check SSC balance of the validator and secures that amount on their own module account.

```[tasklist]
### Tasks
- [ ] https://github.com/thesis/mezo/issues/115
- [ ] https://github.com/thesis/mezo/issues/120
- [ ] https://github.com/thesis/mezo/issues/121
- [ ] https://github.com/thesis/mezo/issues/122
- [ ] https://github.com/thesis/mezo/issues/136
- [ ] https://github.com/thesis/mezo/issues/137
- [ ] https://github.com/thesis/mezo/issues/123
- [ ] https://github.com/thesis/mezo/issues/124
- [ ] https://github.com/thesis/mezo/issues/125
- [ ] https://github.com/thesis/mezo/issues/126
- [ ] https://github.com/thesis/mezo/issues/127
- [ ] https://github.com/thesis/mezo/issues/128
- [ ] https://github.com/thesis/mezo/issues/129
- [ ] https://github.com/thesis/mezo/issues/130
```
